File: Backend/src/main/java/com/bakery/entity/Product.java
package com.bakery.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "products")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank(message = "Product name is required")
    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotNull(message = "Price is required")
    @DecimalMin(value = "0.01", message = "Price must be greater than 0")
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    @DecimalMin(value = "0.00", message = "Cost price cannot be negative")
    @Column(precision = 10, scale = 2)
    private BigDecimal costPrice;

    @Column(length = 500)
    private String imageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProductCategory category;

    @Column(nullable = false)
    @Builder.Default
    private Boolean isAvailable = true;

    @Min(value = 0, message = "Stock quantity cannot be negative")
    @Column(nullable = false)
    @Builder.Default
    private Integer stockQuantity = 0;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (isAvailable == null) {
            isAvailable = true;
        }
        if (stockQuantity == null) {
            stockQuantity = 0;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: Frontend/src/types/product.ts
export type ProductCategory = 'BREAD' | 'BUN' | 'CAKES' | 'PASTRIES' | 'BEVERAGES' | 'COOKIES' | 'OTHER';

export interface Product {
  id: number;
  name: string;
  description: string | null;
  price: number;
  costPrice: number | null;
  category: ProductCategory;
  imageUrl: string | null;
  isAvailable: boolean;
  stockQuantity: number;
}

export interface CreateProductRequest {
  name: string;
  description?: string;
  price: number;
  costPrice?: number;
  category: ProductCategory;
  imageUrl?: string;
  stockQuantity?: number;
}

export interface UpdateProductRequest {
  name: string;
  description?: string;
  price: number;
  costPrice?: number;
  category: ProductCategory;
  imageUrl?: string;
  stockQuantity?: number;
}

export interface StockUpdateRequest {
  quantity: number;
}

export interface AvailabilityUpdateRequest {
  isAvailable: boolean;
}

export const PRODUCT_CATEGORIES: { value: ProductCategory; label: string }[] = [
  { value: 'BREAD', label: 'Bread' },
  { value: 'BUN', label: 'Bun' },
  { value: 'CAKES', label: 'Cakes' },
  { value: 'PASTRIES', label: 'Pastries' },
  { value: 'BEVERAGES', label: 'Beverages' },
  { value: 'COOKIES', label: 'Cookies' },
  { value: 'OTHER', label: 'Other' },
];
